supervisor/shared/serial.c

```diff
@@ -32,26 +32,72 @@
 #include "shared-bindings/terminalio/Terminal.h"
 #include "supervisor/serial.h"
 #include "supervisor/usb.h"
+#include "shared-bindings/microcontroller/Pin.h"
 
 #include "tusb.h"
 
+#if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
+#include "shared-bindings/busio/UART.h"
+busio_uart_obj_t debug_uart;
+byte buf_array[64];
+ringbuf_t rbuf;
+#endif
+
+void serial_early_init(void) {
+#if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
+    rbuf = (ringbuf_t){ buf_array, sizeof(buf_array) };
+    assert_pin_free(DEBUG_UART_TX);
+    assert_pin_free(DEBUG_UART_RX);
+
+    debug_uart.base.type = &busio_uart_type;
+
+    const mcu_pin_obj_t* rx = MP_OBJ_TO_PTR(DEBUG_UART_RX);
+    const mcu_pin_obj_t* tx = MP_OBJ_TO_PTR(DEBUG_UART_TX);
+
+    common_hal_busio_uart_construct(&debug_uart, tx, rx, NULL, NULL, NULL, false,
+                                    115200, 8, PARITY_NONE, 1, 1.0f, 64, &rbuf);
+    common_hal_busio_uart_never_reset(&debug_uart);
+#endif
+}
+
 void serial_init(void) {
     usb_init();
 }
 
 bool serial_connected(void) {
+#if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
+    return true;
+#else
     return tud_cdc_connected();
+#endif
 }
 
 char serial_read(void) {
+#if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
+    if (tud_cdc_connected() && tud_cdc_available() > 0) {
+        return (char) tud_cdc_read_char();
+    }
+    int uart_errcode;
+    char text;
+    common_hal_busio_uart_read(&debug_uart, (uint8_t*) &text, 1, &uart_errcode);
+    return text;
+#else
     return (char) tud_cdc_read_char();
+#endif
 }
 
 bool serial_bytes_available(void) {
+#if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
+    return common_hal_busio_uart_rx_characters_available(&debug_uart) || (tud_cdc_available() > 0);
+#else
     return tud_cdc_available() > 0;
+#endif
 }
 
 void serial_write_substring(const char* text, uint32_t length) {
+    if (length == 0) { // TODO(mark) this is needed because of 'lib/mp-readline/readline.c:71'
```

Comment:
Please include an explanation and not just a link to other code. (The other code may change and move the relevant line.)

Reply:
It was temporary...as in I would fix that before this PR is merged.

Do you know why '\b\b\b\b' is output in headline.c? I don't understand that at all.

Reply:
I don't know why you are seeing '\b's.

Reply:
It's some sort of backspace character...are you happy with having this...if I fix the TODO message?

Reply:
Ya, this is fine. It doesn't make sense to write a string of 0 length.